packages/structures/src/Structure.ts

```diff
@@ -0,0 +1,139 @@
+import { kClone, kData, kMixinConstruct, kMixinToJSON, kPatch } from './utils/symbols.js';
+import type { ReplaceOmittedWithUnknown } from './utils/types.js';
+
+export const DataTemplatePropertyName = 'DataTemplate';
+export const OptimizeDataPropertyName = 'optimizeData';
+
+/**
+ * Represents a data model from the Discord API
+ *
+ * @privateRemarks
+ * Explanation of the type complexity surround Structure:
+ *
+ * There are two layers of Omitted generics, one here, which allows omitting things at the library level so we do not accidentally
+ * access them, in addition to whatever the user does at the layer above.
+ *
+ * The second layer, in the exported structure is effectively a type cast that allows the getters types to match whatever data template is used
+ *
+ * In order to safely set and access this data, the constructor and patch take data as "partial" and forcibly assigns it to kData. To accommodate this,
+ * kData stores properties as `unknown` when it is omitted, which allows accessing the property in getters even when it may not actually be present.
+ * This is the most technically correct way of representing the value, especially since there is no way to guarantee runtime matches the "type cast."
+ */
+export abstract class Structure<DataType extends {}, Omitted extends keyof DataType | '' = ''> {
+	/**
+	 * A construct function used when mixing to allow mixins to set optimized property defaults
+	 *
+	 * @internal
+	 * @remarks This should only be used to set defaults, setting optimized values should be done
+	 * in the mixins `optimizeData` method, which will be called automatically.
+	 * @param data - The full API data received by the Structure
+	 */
+	protected [kMixinConstruct]?(data: Partial<DataType>): void;
+
+	/**
+	 * A function used when mixing to allow mixins to add properties to the result of toJSON
+	 *
+	 * @internal
+	 * @remarks This should only be used to add properties that the mixin optimizes, if the raw
+	 * JSON data is unchanged the property will already be returned.
+	 * @param data - The result of the base class toJSON Structure before it gets returned
+	 */
+	protected [kMixinToJSON]?(data: Partial<DataType>): void;
+
+	/**
+	 * The template used for removing data from the raw data stored for each Structure.
+	 *
+	 * @remarks This template should be overridden in all subclasses to provide more accurate type information.
+	 * The template in the base {@link Structure} class will have no effect on most subclasses for this reason.
+	 */
+	protected static readonly DataTemplate: Record<string, unknown> = {};
```

Comment:
Symbol's not wanted but we should use `DataTemplatePropertyName` / `OptimizeDataPropertyName` when defining this field, lest we end up renaming it at some point and miss this 🙃

Reply:
1 is not applicable and 2 is fine because the override flag would throw an error if we ever rename the field... I don't understand whats wrong with that?

Reply:
I just find it weird to have users extend and override a field but expect them to write it differently than how it is written in the base class property they extend/override.

Reply:
Maybe the added tests suffice to ensure what you're worried about instead?